kube-runtime/src/cache/mod.rs

```diff
@@ -25,6 +29,87 @@ where
     stream.inspect_ok(move |event| store.apply_watcher_event(event))
 }
 
+/// A simple reflector cache around a store and an owned watcher
+pub struct Cache<K>
+where
+    K: Clone + Resource + Send + Sync + 'static,
+    K::DynamicType: Eq + Hash,
+{
+    reader: Store<K>,
+    cache: BoxStream<'static, std::result::Result<watcher::Event<K>, watcher::Error>>,
+}
+
+impl<K> Cache<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+    K::DynamicType: Eq + Hash + Clone + Default,
+{
+    /// Create a Cache on a reflector on a type `K`
+    ///
+    /// Takes an [`Api`] object that determines how the `Cache` listens for changes to the `K`.
+    ///
+    /// The [`ListParams`] controls to the possible subset of objects of `K` that you want to cache.
+    /// For the full set of objects `K` in the given `Api` scope, you can use [`ListParams::default`].
+    #[must_use]
+    pub fn new(api: Api<K>, lp: ListParams) -> Self {
+        Self::new_with(api, lp, Default::default())
+    }
+}
+
+impl<K> Cache<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+    K::DynamicType: Eq + Hash + Clone,
+{
+    /// Create a Cache on a reflector on a type `K`
+    ///
+    /// Takes an [`Api`] object that determines how the `Cache` listens for changes to the `K`.
+    ///
+    /// The [`ListParams`] controls to the possible subset of objects of `K` that you want to cache.
+    /// For the full set of objects `K` in the given `Api` scope, you can use [`ListParams::default`].
+    ///
+    /// This variant constructor is for [`dynamic`] types found through discovery. Prefer [`Cache::new`] for static types.
+    #[must_use]
+    pub fn new_with(api: Api<K>, lp: ListParams, dyntype: K::DynamicType) -> Self {
+        let writer = Writer::<K>::new(dyntype);
+        let reader = writer.as_reader();
+        let cache = reflector(writer, watcher(api, lp)).boxed();
+        Self { reader, cache }
+    }
+
+    /// Retrieve a reading copy of the store before starting the cache
+    #[must_use]
+    pub fn store(&self) -> Store<K> {
+        self.reader.clone()
+    }
+
+    /// Consume the stream and return a future that will run the duration of the program
+    ///
+    /// This should be awaited forever.
+    #[must_use]
+    pub fn run(self) -> BoxFuture<'static, ()> {
```

Comment:
Why is this `Future` boxed rather than just making it an `async fn` (or returning an `impl Future`)?

Reply:
Is this actually something meaningful that users want? Using the cache without listening for updates is typically indicative of a bug IME.

Reply:
Most of the time I use a reflector directly, I just want an in-memory, eventually consistent db and not the underlying events. Generally, this would be for something akin to [version-rs](https://github.com/kube-rs/version-rs/blob/master/version.rs), i.e. apps that gets their state from kubernetes properties on some data type.

Got one that is just responsible for tracking annotations on pods and doing something unrelated that just happens to need the values of those annotations.

Reply:
Fixed pointless boxing in https://github.com/kube-rs/kube-rs/pull/698/commits/6220526db9ede487cacec55d4efc3080202cadad . Thanks!

Reply:
Is the main concern here that we don't catch errors by doing this? I realised that if there are legit errors then throwing it away here is equally problematic. I have changed this `.run()` method to return the first error event from the watcher in https://github.com/kube-rs/kube-rs/pull/698/commits/d3473acebb804effd3fc07b5c852430ef16b095f . Do you think that's enough?

Reply:
Curious about the use-cases here, are any of those uses FOSS, or are you able to go into deeper detail on them?

> Got one that is just responsible for tracking annotations on pods and doing something unrelated that just happens to need the values of those annotations.

Wouldn't that typically mean that you'd want the upstream to update when those annotations change?

Reply:
That aside, even when I have wanted an eventually consistent db for one-off reads (in my case, when writing a load balancer) I've usually found `Store` a pretty poor fit and built my own "db" on top, since `Store` doesn't allow any kind of indexing other than by-name.

Reply:
> Wouldn't that typically mean that you'd want the upstream to update when those annotations change?

The app in question is handling to alertmanager webhooks, and uses pod annotations as a source for who owns the app for incoming alerts. So there's nothing to do inside kubernetes with the information.

`Store` could definitely benefit from ways to find entries, but the `.get(objectref)` method works well if you know the name + namespace. Plus you can make the memory usage pretty small if you write your own serialization struct that discards properties you don't care about.